Extend LanServers to all addresses

krono opened the pull request:
Adds ways to send multicast packages (a) for IPv6 addresses and (b) for all local interfaces.

As it happens, "just" sending a multicast packge will, by default make it's "source" (almost always) the primary IP of the machine. But if the server is listening on multiple (or even all) interfaces, not eveyone will get the memo. So we iterate all local interfaces (that can do multicast and are actually up, but ignore localhost, because what's the point?) and send messages _from_ these addresses. This is a bit involved due to how interfaces/addresses/sockets work in Java.

krono committed:
Extend LanServers to all addresses

Adds ways to send multicast packages (a) for IPv6 addresses and
(b) for all local interfaces.

As it happens, "just" sending a multicast packge will, by default
make it's "source" (almost always) the primary IP of the machine.
But if the server is listening on multiple (or even all) interfaces,
not eveyone will get the memo. So we iterate all local interfaces
(that can do multicast and are actually up, but ignore localhost,
because what's the point?) and send messages _from_ these addresses.
This is a bit involved due to how interfaces/addresses/sockets
work in Java.

Redstonecrafter0 commented on README.md:
```diff
@@ -1,17 +1,20 @@
 # LanServers
 
-This is a small plugin written in Kotlin that runs on all major Minecraft Servers.  
+This is a small plugin that runs on all major Minecraft Servers.  
```
Please to not change anything unrelated to the PR. I'll let the removal of the Kotlin usage pass but format changes are not retaining the projects code style and not topic of the PR.

krono replied:
If you mean.

Redstonecrafter0 commented on pom.xml:
```diff
@@ -6,7 +6,7 @@
     <artifactId>lanservers</artifactId>
     <groupId>net.redstonecraft</groupId>
-    <version>1.1.1</version>
+    <version>2.0.0</version>
```
A major version bump is extreme for this kind of change. 1.2.0 would be better.

krono replied:
Done

Redstonecrafter0 commented on src/main/java/net/redstonecraft/lanservers/LanServersBukkit.java:
```diff
@@ -14,7 +14,7 @@ public void onEnable() {
             try {
                 lanServers = new LanServers(this);
-            } catch (UnknownHostException | SocketException e) {
+            } catch (IOException e) {
```
This is unrelated to the PR and a coding style change. Please check manually before committing AI assisted code. This repeats throughout the PR.

krono replied:
No, this is related, as the constructor can throw an `IOException` in `socket.joinGroup`. This propagates.

Redstonecrafter0 replied:
This is now no longer the case.

Redstonecrafter0 reviewed (changes requested):
Overall two great ideas but b) I still need to test further and a) is not available on the vanilla client as long as [MC-92923](https://bugs.mojang.com/browse/MC/issues/MC-92923) is an open issue. `ff75:230::60` is a small community established multicast address and therefore non-standard. I'm not completely against it tho but it must be opt-in rather than by default on with no opt-out. Also for b) we would need a way to select which local interfaces to use. You may use java properties called dev.redstones.lanservers.interfaces for a comma separated list of addresses, both IPv4 and IPv6. I think this would be the most intuitive way without making a config file to keep the plugin small and simple.

Redstonecrafter0 commented:
Anyway you did a great job and thanks for the PR. Sorry i took a long time to check my GitHub. Keep up the good work.

krono commented:
Ah I didn't fully connect the IPv6 stuff  not being official.

krono commented:
Same Here, sorry I was on vacation.

krono committed:
remove unofficial IPv6 multicast.

Makes everything simpler, too

krono committed:
v2 is too far ahead

krono committed:
fix slip

Redstonecrafter0 commented on README.md:
```diff
@@ -1,17 +1,20 @@
 The only feature is listing the server as a LanServer in the Multiplayer-Menu of Minecraft.  
 
 # Supported Servers
+
```
This is an empty line for no apparent reason.

krono replied:
Some Markdown implementations require empty line after a heading.

Redstonecrafter0 replied:
This would be topic for a separate PR. Besides that i can't find a mention of that in the GFM spec.

krono replied:
It's a best practise: https://www.markdownguide.org/basic-syntax/#heading-best-practices

krono replied:
do you want me to remove those blank lines again?

Redstonecrafter0 replied:
Yes. As I said this would be a different PR with it's own discussion.

Redstonecrafter0 commented on README.md:
```diff
@@ -1,17 +1,20 @@
 - Velocity 3.0.1+
 
 # What does it do exactly?
+
```
This is an empty line for no apparent reason.

krono replied:
see above

Redstonecrafter0 commented on README.md:
```diff
@@ -1,17 +1,20 @@
-`224.0.2.60` is a Multicast Address which means that everyone in you local network that wants
-to get messages sent to `224.0.2.60` will get them like your Minecraft Client when you're in the Multiplayer Menu.
+
+Every 1.5 seconds, the plugin sends a special message with the server port and MOTD to `224.0.2.60` — or `ff75:230::60` for IPv6 — on port 4445 from every interface.
```
The code no longer allows for any IPv6 option so this is no longer true.

krono replied:
sorry, will remove that

Redstonecrafter0 commented on RELEASE_NOTE.md:
```diff
@@ -6,7 +6,10 @@ This is a new version of LanServers.
 
 To use it you need at least Java 8.
 
+It supports IPv4 and IPv6
```
No longer true.

krono replied:
fixed

Redstonecrafter0 commented on RELEASE_NOTE.md:
```diff
@@ -6,7 +6,10 @@ This is a new version of LanServers.
+It supports IPv4 and IPv6
+
 # Compatibility
+
```
This is an empty line for no apparent reason.

Redstonecrafter0 commented on src/main/java/net/redstonecraft/lanservers/LanServers.java:
```diff
@@ -3,19 +3,26 @@
-    private final InetAddress multicastAddress = InetAddress.getByName("224.0.2.60");
-    private final int multicastPort = 4445;
-    private final DatagramSocket socket = new DatagramSocket();
+    private final Inet4Address multicastAddress = (Inet4Address) InetAddress.getByName("224.0.2.60");
```
Is there a particular reason for the usage of Inet4Address over InetAddress? Seems like just a codestyle change.

krono replied:
Leftover from Inet6/Inet4 but I think it is more accurate now, since this is IPv4-_only_ multicast

Redstonecrafter0 commented on src/main/java/net/redstonecraft/lanservers/LanServers.java:
```diff
@@ -3,19 +3,26 @@
-    private final int multicastPort = 4445;
-    private final DatagramSocket socket = new DatagramSocket();
+    private final Inet4Address multicastAddress = (Inet4Address) InetAddress.getByName("224.0.2.60");
+    private final static int multicastPort = 4445;
```
Is there a reason for this to be static? Seems like a codestyle change.

krono replied:
It's a constant and used in static `getSourceSockets`

Redstonecrafter0 replied:
Then the continued question would be why getSourceSockets is static? Agree on the constant part tho it should be all uppercase in that case. For this i would tho consider this a change unrelated to the PRs purpose, so maybe for another PR.

krono replied:
Because its initalizing a final field and not in the constructor?

Redstonecrafter0 commented on src/main/java/net/redstonecraft/lanservers/LanServers.java:
```diff
@@ -3,19 +3,26 @@
-    public LanServers(ILanServersPlugin plugin) throws UnknownHostException, SocketException {
+    public LanServers(ILanServersPlugin plugin) throws IOException {
         this.plugin = plugin;
+        for (MulticastSocket socket : sources) {
```
This should probably be moved to enable(). That would allow the plugin to be reloaded or managed by other plugins. Also this would revert the exception signature and the error should be handled in enable().

krono replied:
Ok, will do.

Redstonecrafter0 replied:
The enable methods signature was changed to throws IOException. Instead the error should be handled within the method.

krono replied:
No it should not. It is at the same level of abstraction as `UnknownHostException, SocketException` in the constructor.  It is quite apt to expect this to succeed in `enable` and thorw if not.

Redstonecrafter0 commented on src/main/java/net/redstonecraft/lanservers/LanServers.java:
```diff
@@ -24,18 +31,55 @@ public void enable() {
 
+    private static List<MulticastSocket> getSourceSockets() {
+        try {
+            // Take all network interfaces
```
All comments in this file describe pretty apparent code behavior. Probably just some left overs.

krono replied:
Hmm, I tried to be courteous to the reader, but yeah, canr emove that

Redstonecrafter0 commented on src/main/java/net/redstonecraft/lanservers/LanServersBukkit.java:
```diff
@@ -2,8 +2,8 @@
-import java.net.SocketException;
-import java.net.UnknownHostException;
+import java.io.IOException;
+
```
This is an empty line for no apparent reason.

krono replied:
fixed

Redstonecrafter0 commented on src/main/java/net/redstonecraft/lanservers/LanServersVelocity.java:
```diff
@@ -9,13 +9,13 @@
-import java.net.SocketException;
-import java.net.UnknownHostException;
+import java.io.IOException;
+
```
This is an empty line for no apparent reason.

krono replied:
fixed

Redstonecrafter0 commented on src/main/java/net/redstonecraft/lanservers/LanServersVelocity.java:
```diff
@@ -9,13 +9,13 @@
     id = "lanservers",
     name = "LanServers",
-    version = "1.1.0",
+    version = "2.0.0",
```
Here is a version that differs from the pom.xml. Sadly velocity does not easily allow a better way.

krono replied:
ah forgot that, sorry.

Redstonecrafter0 commented on src/main/java/net/redstonecraft/lanservers/LanServers.java:
```diff
@@ -24,18 +31,55 @@ public void enable() {
+                                try {
+                                    return new MulticastSocket(insaddr);
+                                } catch (IOException e) {
+                                    return null;
```
No stacktrace is printed and no comment describing a reason for silencing the error.

krono replied:
I make a comment

Redstonecrafter0 replied:
There is no comment in the catch block describing the reason.

Redstonecrafter0 reviewed (commented):
Mainly a whole bunch of small codestyle concerns. I haven't gotten around yet to fully test the plugin since i haven't set up a VM with multiple NICs yet.

krono committed:
remove IPv6 mention in readme

krono committed:
Remove IPv6 mention from release notes

krono committed:
address review concerns

krono committed:
comment uncommon try-catch-usage

krono committed:
move multicast-group join to enable.

this effects, that the IOException moves to the enable method
and must be caught in the impls

krono committed:
typo

Redstonecrafter0 reviewed (commented):
Please try to keep the changes minimal and to the topic.

krono commented:
Look, this is a ~200 Lines-of-Code simple project. I wanted to contribute. 
I did not expect that kind of spanish inquisiton. 

Feel free to close if you dont want it.

Redstonecrafter0 commented:
Seems abandoned.
a) is now being tracked in #5 and b) is now implemented and expanded upon as part of #4

krono commented:
Abandoned is not the right word.
Aber jeder wie er will.

Redstonecrafter0 commented:
Please stick to english so anybody in the future can still easily understand what was written.

The combination of the 10 days of the non-addressed change request combined with the comment
> Look, this is a ~200 Lines-of-Code simple project. I wanted to contribute. I did not expect that kind of spanish inquisiton.
> 
> Feel free to close if you dont want it.

led me to believe the PR was abandoned.  
If you feel like there is a better word, feel free to share it.

If you have anything to add regarding the topics of the now closed PR you can do so at #4 and #5.
If you want to discuss something else you can now do so in the discussions tab.

krono commented:
no.
